core/trino-main/src/main/java/io/trino/server/security/oauth2/OidcDiscovery.java

```diff
@@ -0,0 +1,134 @@
+    private final Optional<String> authUrl;
+    private final Optional<String> tokenUrl;
+    private final Optional<String> jwksUrl;
+    private final Optional<String> userinfoUrl;
+    private final NimbusHttpClient httpClient;
+
+    @Inject
+    public OidcDiscovery(OAuth2Config oauthConfig, OidcDiscoveryConfig oidcConfig, NimbusHttpClient httpClient)
+    {
+        requireNonNull(oauthConfig, "oauthConfig is null");
+        issuer = new Issuer(requireNonNull(oauthConfig.getIssuer(), "issuer is null"));
+        requireNonNull(oidcConfig, "oidcConfig is null");
+        userInfoEndpointEnabled = oidcConfig.isUserInfoEndpointEnabled();
+        discoveryTimeout = Duration.ofMillis(requireNonNull(oidcConfig.getDiscoveryTimeout(), "discoveryTimeout is null").toMillis());
+        accessTokenIssuer = requireNonNull(oidcConfig.getAccessTokenIssuer(), "accessTokenIssuer is null");
+        authUrl = requireNonNull(oidcConfig.getAuthUrl(), "authUrl is null");
+        tokenUrl = requireNonNull(oidcConfig.getTokenUrl(), "tokenUrl is null");
+        jwksUrl = requireNonNull(oidcConfig.getJwksUrl(), "jwksUrl is null");
+        userinfoUrl = requireNonNull(oidcConfig.getUserinfoUrl(), "userinfoUrl is null");
+        this.httpClient = requireNonNull(httpClient, "httpClient is null");
+    }
+
+    @Override
+    public OAuth2ServerConfiguration get()
+    {
+        return Failsafe.with(new RetryPolicy<>()
+                        .withMaxAttempts(-1)
+                        .withMaxDuration(discoveryTimeout)
+                        .withDelay(Duration.ofSeconds(1))
+                        .abortOn(IllegalStateException.class)
+                        .onFailedAttempt(attempt -> LOG.debug("OpenID Connect Metadata read failed: %s", attempt.getLastFailure())))
+                .get(() ->
+                        httpClient.execute(new OIDCProviderConfigurationRequest(issuer), response -> {
+                            int statusCode = response.getStatusCode();
+                            if (statusCode != OK.code()) {
+                                // stop on any client errors other than REQUEST_TIMEOUT and TOO_MANY_REQUESTS
+                                if (statusCode < 400 || statusCode >= 500 || statusCode == REQUEST_TIMEOUT.code() || statusCode == TOO_MANY_REQUESTS.code()) {
+                                    throw new RuntimeException("Invalid response from OpenID Metadata endpoint: " + statusCode);
+                                }
+                                else {
+                                    throw new IllegalStateException(format("Invalid response from OpenID Metadata endpoint. Expected response code to be %s, but was %s", OK.code(), statusCode));
+                                }
+                            }
+                            JSONObject body = response.getContentAsJSONObject();
+
+                            OIDCProviderMetadata metadata = OIDCProviderMetadata.parse(body);
```

Comment:
if you call `response.getContent()` (which returns `String`) you can pass it to `OIDCProviderMetadata.parse(String)` this eliminates the need for the extra JSON library. 

Also `body` can just be inlined, the variable isn't used anywhere else.

Reply:
`JSONObject` was used to get `access_token_issuer` which is not part of the spec and is not recognized by `OIDCProviderMetadata.parse`. That said, I've replaced this usage with jackson and removed the dependency to this Json library.